proxy/server.go

```diff
@@ -18,74 +18,135 @@
 package proxy
 
 import (
-	"bufio"
-	"fmt"
+	"io"
+	"log"
 	"net"
-	"net/http"
+	"net/http/httputil"
 	"strings"
 
-	"github.com/elazarl/goproxy"
+	"github.com/inconshreveable/go-vhost"
+	"github.com/pkg/errors"
+	"github.com/soheilhy/cmux"
 	netproxy "golang.org/x/net/proxy"
 )
 
-type stickyHasher interface {
-	Hash(ip string) string
-}
-
 type domainTracker interface {
 	Inc(domain string)
 }
 
+type proxyServer struct {
+	addr   string
+	dialer netproxy.Dialer
+	sm     *stickyMapper
+	dt     domainTracker
+}
+
 // NewServer returns new instance of HTTP transparent proxy server
-func NewServer(addr string, upstreamDialer netproxy.Dialer, mapper stickyHasher, dt domainTracker) *http.Server {
-	server := goproxy.NewProxyHttpServer()
-	server.Verbose = true
-	server.NonproxyHandler = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		if req.Host == "" {
-			fmt.Fprintln(w, "Cannot handle requests without Host header, e.g., HTTP 1.0")
-			return
-		}
-		req.URL.Scheme = "http"
-		req.URL.Host = req.Host
-		server.ServeHTTP(w, req)
-	})
-
-	server.ConnectDial = upstreamDialer.Dial
-	server.OnRequest().HandleConnectFunc(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
-		return goproxy.OkConnect, host
-	})
-
-	server.OnRequest().DoFunc(func(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
```

Comment:
Goproxy was additionally serving 3rd endpoint - explicitly accepting CONNECT queries, which allowed to test proxying logic wo configured tunnel redirect with iptables.

It's documented here https://github.com/mysteriumnetwork/openvpn-forwarder

```
3. Check if forwarder is redirecting requests to upstream HTTPS proxy
FORWARDER_IP=`docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' forwarder`
curl --proxy $FORWARDER_IP:8080 "http://ipinfo.io"
```

Reply:
Added support for direct requests too.